### ISSUE TITLE
[Bug Report] v-menu: item in activator slot disappears when pushing route

### ISSUE THREAD

**Issue by lukef** [vuetifyjs/vuetify#10578]
EDIT: repro https://github.com/vuetifyjs/vuetify/issues/10578#issuecomment-612163991

### Environment
**Vuetify Version:** 2.2.12
**Vue Version:** 2.6.11
**Browsers:** Chrome 80.0.3987.100
**OS:** Linux x86_64

### Steps to reproduce
1. Add a `v-menu` like the following:

```
    <v-menu
      ref="date-picker"
      v-model="showingPicker"
      :close-on-content-click="false"
      offset-y
      max-width="290px"
      min-width="290px"
    >
      <template v-slot:activator="{ on }">
        <v-text-field
          v-model="dateFormatted"
          hint="MM/DD/YYYY format"
          persistent-hint
          :outlined="true"
          append-icon="mdi-calendar"
          @blur="updatedInputDate()"
          v-on="on"
        ></v-text-field>
      </template>
      <v-date-picker v-model="date" no-title @input="selectedCalendarDate()"></v-date-picker>
    </v-menu>
```

2. Navigate to another view via: `this.$router.push('/somewhere')`.
3. The `v-text-field` element will disappear

### Expected Behavior
The `v-text-field` element should not disappear

### Actual Behavior
The `v-text-field` element will disappear

### Other comments
Also, is it possible attach `v-on` to the `@click:append-icon` event?

**Comment by lukef** [comment]
I'll try to get a minimal reproduction up soon

**Comment by jacekkarczmarczyk** [comment]
This issue is being closed due to:

- [x] Reproduction environment not provided (inactivity for 14 days)
- [ ] Requested information not provided (inactivity for 14 days)
- [ ] Violation of the [Code of Conduct](https://github.com/vuetifyjs/vuetify/blob/dev/CODE_OF_CONDUCT.md)

**Comment by zeroarst** [comment]
I am having this issue. Whenever route changed, the element in <template v-slot:activator="{ on }">...</> disappears..

**Comment by dannielhugo** [comment]
I am having this issue too. I've tried to reproduce on Codepen but It was unsuccessful :disappointed: 
@lukef Have you succeeded with this issue reproduction?

**Comment by lukef** [comment]
We actually ended up swapping this component out but I will try to create a vanilla reproduction case this week. That said, I can tell you it is still happening in our code if I add the control back.

**Comment by nchan7** [comment]
@lukef were you able to recreate the issue? I am running into the same issue with the Timepicker component

**Comment by timaxxer** [comment]
I have actually have the same issue, All component wrapped in `<template v-slot:activator="{ on }">...</>` disappear before the route is unmounted.

**Comment by eikooc** [comment]
I think I found out a situation where it is happening, and how to solve it. At least in my case.

When the structure inside the `<template v-slot:activator="{ on }"> ... </template>` changes. The rendering breaks. So if you have a `v-if` inside that template that switches between different elements, it breaks. 

Can somebody confirm this?

**Comment by timaxxer** [comment]
I tried to reproduce it and I found that I am not able to reproduce it only with Vuetify and vue-router. But I was able to reproduce it with nuxt. I dont know why within nuxt the problem is present.

I prepared a simple codesandbox to show this.

https://w63ms.sse.codesandbox.io/

Scenario:
By navigation from /about to home page, The dropdown button dissapear until the Home page is rendered completely.

@eikooc this is not look like to be related to structure inside template v-slot

@jacekkarczmarczyk Is it possible to investigate and maybe reopen the issue?

**Comment by jacekkarczmarczyk** [comment]
Your repro is not something we can debug (no source) @timaxxer

**Comment by timaxxer** [comment]
Sorry here is the link
https://codesandbox.io/s/vibrant-hermann-w63ms

**Comment by timaxxer** [comment]
The problem is related to the transitions feature on the router in nuxt.

By disabling the transitions feature in `nuxt.config`, the problem disappear. So hope this clue will help you find a solution to be able to enable transition on router and fix the problem in vuetify.
```
...
  features: {
    transitions: false
  },
...
```

**Comment by tiboradam** [comment]
Not 100% sure I'm using Vuetify's btn/menu for a confirmation "dialog" right, but here's my implementation and I get the same issue. There's another issue with the activator's width (in this case the button) being larger than the menu's. But that's not functional, right? :)

I think it's related to the variable changing on the activator v-if.
I've tried some simple combination. Feel free to experiment. 

Made this codepen, hope it helps:
https://codepen.io/tiboradam/pen/qBOjMor

**Comment by tiboradam** [comment]
Can anyone suggest a solution to this? I'd be happy with a temporary, hacky one too. :)

**Comment by denis-ilchishin** [comment]
I have the same issue using pure Vue (not Nuxt). But not only with v-menu, but any component that has activator slot, such as, for example, v-tooltip. Everything within activator slot disappears before current route is unmounted. Unfortunately, I can't reproduce that on codepen as well.

**Comment by td-kallmon** [comment]
@tiboradam I tried @timaxxer 's fix (just above your post) and it worked for me.

**Comment by tiboradam** [comment]
Yeah, but I'm not using Nuxt. Where should I put that feature configuration?

**Comment by td-kallmon** [comment]
Yeah, I noticed after I commented that some others weren't using Nuxt, making that solution only workable. So I deleted my comment. Not fast enough :p

It's interesting that this setting in Nuxt will fix it, but there's no obvious way to change such a setting in Vuetify, at least not that I could find.

**Comment by jarkt** [comment]
Do you mean that disable transitions will fix it? It fixes nothing but due to an animation time of 0s it maybe doesn't appear. I use transitions, so it's not a solution for me. Or what was the mysterious deleted post?

**Comment by td-kallmon** [comment]
@jarkt You're absolutely right, it's not a fix, it's a bit of a hack. My original response was to someone who said they were willing to accept even a hack, so I pointed to an earlier comment I thought they might not have seen. But... well, yeah. My mistake, all around.

**Comment by augustinoMageka** [comment]
I added 
{ 
   features: { 
         transitions: false;
    }
}
to my nuxt.config.js and yey it worked.

**Comment by tiboradam** [comment]
Good for ya. :)
What about in a Vue stack that doesn't have Nuxt.. like in my examples?

**Comment by mantlee** [comment]
looking for a fix as well without nuxt

**Comment by yoldar** [comment]
I guess it is something related to the "transition". I got <transition name="fade"> above router-view:
```
vue:
<template lang="pug">
    v-app
      transition(name="fade")
          router-view

stylus:
// Fade annimation
.fade-enter-active, .fade-leave-active
  transition opacity .3s
.fade-enter, .fade-leave-to
  opacity 0
```

I increased fade animation to show you how menu disappears  
![2020-08-01 01 05 21](https://user-images.githubusercontent.com/1597578/89068580-5a7cc100-d393-11ea-812b-6dec89ca8ef0.gif)

**Comment by Thomas-lhuillier** [comment]
I had pretty much the same issue, I solved it by adding the `attach` prop on my `v-menu`.

```vue
<v-menu
  bottom
  left
  origin="top right"
  attach="">
```

Maybe it can also help here :)

**Comment by mwgamble** [comment]
I'm seeing this issue with tooltip activators. It only happens when changing routes. I am using nuxt, and don't have the luxury of being able to turn off transitions.

**Comment by JulienAramino** [comment]
I have the same issue with v-btn in v-dialog with nuxt using 
`template(v-slot:activator="{ on, attrs }"), `
is there a fix  better than 
```
features: {
transitions: false;
}
```

**Comment by eduardo76** [comment]
This problem is related to transitions aplied to routes and the use of **v-slot:activador**. 
In my projects I use a app-bar with some buttons with tooltips.
Instead of using **v-slot:activator**, I have used the props **activator** in tooltips.
Example:
**v-slot:activator**
```vue
<v-tooltip bottom>   
  <template v-slot:activator="{ on, attrs }">  
    <v-btn color="primary" dark v-bind="attrs" v-on="on">
      Button
    </v-btn>
  </template>
  <span>Tooltip</span>
</v-tooltip>
```

**Props activator**
```vue
<v-btn id="bt_button" color="primary" dark>
  Button
  <v-tooltip bottom activator="#bt_button">
    <span>Tooltip</span>
  </v-tooltip>
</v-btn>
```

This problem can be seen in many others v-slot:activator, including v-menu.

**Comment by yoldar** [comment]
> This problem is related to transitions aplied to routes and the use of **v-slot:activador**.
> In my projects I use a app-bar with some buttons with tooltips.
> Instead of using **v-slot:activator**, I have used the props **activator** in tooltips.
> Example:
> **v-slot:activator**
> 
> ```
> <v-tooltip bottom>   
>   <template v-slot:activator="{ on, attrs }">  
>     <v-btn color="primary" dark v-bind="attrs" v-on="on">
>       Button
>     </v-btn>
>   </template>
>   <span>Tooltip</span>
> </v-tooltip>
> ```
> 
> **Props activator**
> 
> ```
> <v-btn id="bt_button" color="primary" dark>
>   Button
>   <v-tooltip bottom activator="#bt_button">
>     <span>Tooltip</span>
>   </v-tooltip>
> </v-btn>
> ```
> 
> This problem can be seen in many others v-slot:activator, including v-menu.

Thanks! This solved problem!

**Comment by gangsthub** [comment]
As @eikooc [commented](https://github.com/vuetifyjs/vuetify/issues/10578#issuecomment-611438609) I had this issue because the only child of `<template v-slot:activator="{ on }"></template>` had a `v-if` that was toggling it's visibility.

I solved it by wrapping the "activator button" in a `div` that is not bound to the `v-if`:

```vue
<v-menu
  v-model="isMenuOpen"
  v-bind="{...otherAttrs}"
>
  <template v-slot:activator="{ on }">
    <div>
      <my-button
        v-if="condition"
        @click="isMenuOpen = !isMenuOpen"
        v-on="on"
      >
        Some text
      </my-button>
    </div>
  </template>

  <v-card>Menu content...</v-card>
</v-menu>
```

**Comment by mstralka** [comment]
I followed @eduardo76 advice and am using the activator prop now instead, but why is this necessary? Is there a way to fix this so the activator slot can still be used with transitions?

**Comment by eduardo76** [comment]
> I followed @eduardo76 advice and am using the activator prop now instead, but why is this necessary? Is there a way to fix this so the activator slot can still be used with transitions?

That's a good question.
In some situations my solution is not good. When using inside a table, for instance, is not possible to use **id** and **activator** together. The **id** would have to be dynamicly filled to be different for **activator** to work properlly.

**Comment by bartosz-brodzik** [comment]
What is state of this? It's really annoying and brakes the layout when route changes...

**Comment by eduardo76** [comment]
> What is state of this? It's really annoying and brakes the layout when route changes...

I have found an other solution. If you use some of vuetify transition components, you can add the prop hide-on-leave and it will hide everything before leaving current route.
[https://vuetifyjs.com/en/api/v-slide-y-transition/#hide-on-leave](https://vuetifyjs.com/en/api/v-slide-y-transition/#hide-on-leave)

**Comment by mzduke** [comment]
There are also old issue reports that actually about the same defect but with multiple suggested solutions:
- https://github.com/vuetifyjs/vuetify/issues/7422
- https://github.com/vuetifyjs/vuetify/issues/9276

**Comment by verstoff** [comment]
I am using one of the latest versions of Vuetify with Nuxt.js and having the same problem with components, that uses _Detachable_ mixin (like v-menu):

- The component is invisible on SSR and appears only after hydration.
- The component disappears for a moment when route is changing on the client-side.

Both of these problems look visually like terrible layout shifting.

The component is first rendered into the wrapper (root component's div) and then detached from it and attached to the top, but before that, it is not visible. Most workarounds suggest nesting the activator and component in one item instead of detaching or use different components (use activator prop). But extra nesting doesn't always work. She often breaks the layout, especially since she is not required here. And using a separate activator requires laborious changes to the correct code just for the sake of fixing.

I wrote my very simple workaround, which solves the problem in my case without additional nesting. The idea is to first display only the activator, and then replace it with a real Vuetify component that does detach-attach. Also, there is no need to change the original way of using the component.


```
// YourComponent.vue
<!-- As usual -->
<template>
<v-menu>
    <template v-slot:activator="{ on, attrs }">
      <v-btn v-bind="attrs" v-on="on">
      </v-btn>
    </template>
    <v-list>
	    ...
    </v-list>
  </v-menu>
</template>

<script>
/* You can move this declaration to a separate file or replace v-menu globally somewhere.
This is just a simple Proof-of-concept */

import { VMenu } from 'vuetify/lib/components/VMenu'
const VMenuCustom = {
	components: { VMenu },
	data() {
		return {
			interactive: false
		}
	},
	render(h) {
		if (this.interactive) {
			return h('v-menu', {
				attrs: this.$attrs,
				on: this.$listeners,
				scopedSlots: this.$scopedSlots
			})
		} else {
			return this.$scopedSlots.activator()
		}
	},
	mounted() {
		this.interactive = true
	}
}

export default {
  // Use our wrapper instead original VMenu 
  components: { VMenu: VMenuCustom }
}
</script>

```

**Comment by kanteankit** [comment]
Hi, thank you for the workaround for `v-menu` and `v-tooltip`.
I am facing this same issue with `v-edit-dialog` because it uses v-menu. Does anyone have a workaround for `v-edit-dialog`? It provides a `default` slot which internally [uses the `activator` slot of v-menu](https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VDataTable/VEditDialog.ts#L123)

**Comment by bendytree** [comment]
This should work on all v-menu instances. May need to run on `v-tooltip` as well.

```
Vue.mixin({
  beforeDestroy() {
    if (this.$options.name === 'v-menu') {
      this.activatorNode = null;
    }
  },
});
```

**Comment by fabis94** [comment]
I was having similar issues and I decided to actually debug through Vue rendering code to understand what's going on. It seems like for some reason the `activator` slot of `v-list-group` is treated as a scoped slot and when `normalizeScopedSlots()` is invoked, it's correctly attached to the main $slots property making it available through `$slots.activator`.

But then on subsequent re-renders it skips the `normalizeScopedSlots()` algorithm (because `slots._normalized` is set) and when that happens it no longer attaches the $scopedSlots.activator function to $slots. And then `this.$slots.activator` becomes undefined in the `v-list-group` render function.

On top of this: I checked Vue docs and Vue suggests always using `this.$scopedSlots` in render functions, even when dealing with the basic kind of slots ([source](https://v2.vuejs.org/v2/api/#vm-scopedSlots)):

> If you work with render functions, it is now recommended to always access slots via $scopedSlots, whether they currently use a scope or not. This will not only make future refactors to add a scope simpler, but also ease your eventual migration to Vue 3, where all slots will be functions.

And if you check the source code of `v-list-group` in this repo, the render function only uses `this.$slots`.

**So what is the solution?**

To define the `activator` slot using the old syntax that attaches it to `$slots` not `$scopedSlots`.

So do it like this:
```
          <template slot="activator">
               ...
          </template>
```

Not like this:
```
          <template #activator>
               ...
          </template>
```

**Comment by TiborPL** [comment]
Another kind of hack is to give the v-menu a key property that is changed when the variable is change that makes the activator shown or hidden.

```vue
<v-menu
  :key="pleaseRerender"
  v-model="isMenuOpen"
  v-bind="{...otherAttrs}"
>
  <template v-slot:activator="{ on }">
    <div>
      <v-text-field
        v-if="condition"
        @click="isMenuOpen = !isMenuOpen"
        v-on="on"
      >
        Some text
      </v-text-field>
  </template>
  <v-date-picker v-model="pickedDate" />
</v-menu>
```

then 

```vue
  watch: {
    'someOtherCondition'() {
      this.dueRender++
    }
  }
```

**Comment by eugene-hovorun** [comment]
For me this happened with Nuxt, after build only. The workaround, which helped me, is an external [activator](https://vuetifyjs.com/en/api/v-menu/#props-activator).

**Comment by lfgtavora** [comment]
this was openned 2years ago and i still lost a day trying to figure out what is going one. This solutions worked well! this warning must in documentation
